app/shared/user-agent-client.js

```diff
@@ -0,0 +1,134 @@
+/*
+ Copyright 2016 Mozilla
+
+ Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ this file except in compliance with the License. You may obtain a copy of the
+ License at http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software distributed
+ under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ specific language governing permissions and limitations under the License.
+ */
+
+import path from 'path';
+import EventEmitter from 'events';
+import WebSocket from 'ws';
+import request from 'request';
+import { setTimeout } from 'timers';
+import childProcess from 'child_process';
+import * as endpoints from './constants/endpoints';
+
+function uaRequest(service, options) {
+  return new Promise((resolve, reject) => {
+    request[options.method.toLowerCase()]({
+      url: `${endpoints.UA_SERVICE_HTTP}${service}`,
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+    }, (err, response, body) => {
+      if (err) {
+        reject(err);
+        return;
+      }
+
+      resolve(JSON.parse(body));
+    });
+  });
+}
+
+class UserAgent extends EventEmitter {
+  constructor(ws, clientCount) {
+    super();
+
+    this.ws = ws;
+    this.clientCount = clientCount;
+
+    this.ws.on('message', (data) => {
+      data = JSON.parse(data);
+      const message = data.message;
+      delete data.message;
+      this.emit(message, data);
+    });
+  }
+
+  startSession() {
+    return uaRequest('/session/start', {
+      method: 'POST',
+    });
+  }
+}
+
+const timeout = (ms) => new Promise(resolve => setTimeout(resolve, ms));
+
+function startService() {
```

Comment:
Maybe, it would mean splitting this into two files, one that attempts to connect and starts the UA service if it fails and a second that just assumes it can always connect.